### PR TITLE
 feat: primary type

### PR DESCRIPTION
**Pull request by P0lip**

### CODE REVIEW

File: src/utils/walk.ts [hunk 1]
```diff
@@ -84,7 +82,20 @@ function processNode(node: JSONSchema4): SchemaNode | void {
     } as IRefNode;
   }
 
-  if (combiner !== undefined) {
+  if (combiner === undefined) {
+    const base: IBaseNode = {
+      id: assignId(node),
+      primaryType: getPrimaryType(node),
+      type: node.type || inferType(node),
```
Comment (marbemac) [inline src/utils/walk.ts]: Why have two properties? When will a node ever have a defined type AND a primary type?

Comment (marbemac) [reply src/utils/walk.ts]: Any reason we can't just update `inferType` and be done with it?

Comment (P0lip) [reply src/utils/walk.ts]: `primaryType` - this props drives the rendering logic
`type` -  displayed in UI

File: src/utils/isSchemaViewerEmpty.ts [hunk 2]
```diff
@@ -1,6 +1,5 @@
 import { ISchema } from '@stoplight/types';
-import get = require('lodash/get');
-import isEmpty = require('lodash/isEmpty');
+import { get, isEmpty } from 'lodash';
```
Comment (casserni) [inline src/utils/isSchemaViewerEmpty.ts]: _get and _isEmpty

File: src/utils/walk.ts [hunk 3]
```diff
@@ -33,41 +33,39 @@ function assignNodeSpecificFields(base: IBaseNode, node: JSONSchema4) {
...
 
-  if (node.type !== undefined && combiner === undefined) {
-    const base: IBaseNode = {
-      id: assignId(node),
-      type: getType(node),
-      validations: getValidations(node),
-      annotations: getAnnotations(node),
-      enum: node.enum,
-    };
+function getPrimaryType(node: JSONSchema4) {
```
Comment (casserni) [inline src/utils/walk.ts]: what if type is `["object", "array"]`

File: src/utils/walk.ts (hunk 1 above)
Comment (casserni) [reply src/utils/walk.ts]: Does primary type need to be store on the node if not used in the ui? Can we just call `getPrimaryType` in renderSchema when its needed

File: src/utils/walk.ts (hunk 3 above)
Comment (casserni) [reply src/utils/walk.ts]: Really the user should be using an oneOf like below, but if having both in type is valid we definitely need to support it

```
{
    "oneOf": [
        {"$ref": "#/definitions/singleObject"}, // plain object
        {
            "type": "array", // array of plain objects
            "items": {"$ref": "#/definitions/singleObject"}
        }
    ],
    "definitions": {
        "singleObject": {...}
    }
}```

Comment (casserni) [reply src/utils/walk.ts]: This also means that if type is `["object", "array"]` we need to infer that its a combiner type oneOf so it renders properly in the ui

Comment (P0lip) [reply src/utils/walk.ts]: Yeah, shared the same concern here https://github.com/stoplightio/json-schema-viewer/pull/17#discussion_r283318479.
I'm not sure if displaying `oneOf` is safe though. To me, it sounds like a nice idea, but we'll need to make it reliable (transform the data correctly).
 @marbemac, what's your take on this?

File: src/utils/walk.ts (hunk 1 above)
Comment (P0lip) [reply src/utils/walk.ts]: Fair enough!

File: src/utils/walk.ts (hunk 3 above)
Comment (marbemac) [reply src/utils/walk.ts]: I think we don't worry about this use case until a customer complains about it.

**Pull request merged**